test/boost/loading_cache_test.cc

```diff
@@ -428,6 +429,163 @@ SEASTAR_TEST_CASE(test_loading_cache_eviction_unprivileged) {
+        auto load_len95 = [] (const int& key) { return make_ready_future<sstring>(tests::random::get_sstring(95)); };
+
+        using namespace std::chrono;
+        utils::loading_cache<int, sstring, 1, utils::loading_cache_reload_enabled::no, sstring_length_entry_size> loading_cache(100, 1h, testlog);
+        auto stop_cache_reload = seastar::defer([&loading_cache] { loading_cache.stop().get(); });
+
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 0);
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 0);
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 0);
+
+        loading_cache.get_ptr(1, load_len1).discard_result().get();
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 0);
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 1);
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 1);
+
+        loading_cache.get_ptr(2, load_len5).discard_result().get();
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 0);
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 6);
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 2);
+
+        // Move "2" to privileged section by touching it the second time.
+        loading_cache.get_ptr(2, load_len5).discard_result().get();
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 5);
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 1);
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 2);
+
+        loading_cache.get_ptr(3, load_len10).discard_result().get();
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 5);
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 11);
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 3);
+
+        // Move "1" to privileged section. load_len10 should not get executed, as "1"
+        // is already in the cache.
+        loading_cache.get_ptr(1, load_len10).discard_result().get();
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 6);
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 10);
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 3);
+
+        // Flood cache with elements of size 10,
+        // unprivileged. "1" and "2" should stay in the privileged section.
+        for (int i = 11; i < 30; i++) {
+            loading_cache.get_ptr(i, load_len10).discard_result().get();
+        }
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 6);
+        // We shrink the cache BEFORE adding element,
+        // so after adding the element, the cache
+        // can exceed max_size...
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 100); 
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 12);
```

Comment:
While writing this test, I saw another "problem" introduced in 1a9c6d9fd3af5b9dd06bc51128fdcd3185456c70. That commit also changed that the `shrink()` is called before adding a new entry, not after adding a new entry. (The reasoning was that if we just added an only element to unprivileged section, shrinking the cache would evict just added entry).

The effect of such change is that the size of the cache can get larger than `_max_size` by the last entry added.

Decided not to pollute this PR by fixing too many issues.

Reply:
Don't think it needs to be changed. It's not like _max_size is a perfect size and _max_size+1 will cause the sun to collapse into a black hole - we need some limit, and it's not too bad if the limit is a little fuzzy.

(it's much more interesting to convert this limit from a count to a byte size, since that's what we really want to limit, but also much harder)

Reply:
> (it's much more interesting to convert this limit from a count to a byte size, since that's what we really want to limit, but also much harder)

What do you mean by "this limit"? `_max_size` is the "byte size" ("memory footprint") limit, not number of entries.

This brings me to the next question: maybe we should change the naming we use in `loading_cache`: `_max_size` changed to `_max_memory_footprint` (and others similarly)? In my first attempt, I mistook `_max_size` for number of entires and wrote wrong code. This also leads to a strange code like this - current master:

https://github.com/scylladb/scylla/blob/aee94b7176aad5dfb1b302c1a64b257c6dc4fe7a/utils/loading_cache.hh#L377-L384

`size()` returning number of entires, `memory_footprint()` returns the memory footprint by reading from... `_current_size` and `timestamped_val::size()` returns the memory footprint.

Reply:
> > (it's much more interesting to convert this limit from a count to a byte size, since that's what we really want to limit, but also much harder)
> 
> What do you mean by "this limit"? `_max_size` is the "byte size" ("memory footprint") limit, not number of entries.
> 

I meant _max_size, I incorrectly assumed it was an entry count limit.

> This brings me to the next question: maybe we should change the naming we use in `loading_cache`: `_max_size` changed to `_max_memory_footprint` (and others similarly)? In my first attempt, I mistook `_max_size` for number of entires and wrote wrong code.

Agree. _list.size() really smells like an element count.

> This also leads to a strange code like this - current master:
> 
> https://github.com/scylladb/scylla/blob/aee94b7176aad5dfb1b302c1a64b257c6dc4fe7a/utils/loading_cache.hh#L377-L384
> 
> `size()` returning number of entires, `memory_footprint()` returns the memory footprint by reading from... `_current_size` and `timestamped_val::size()` returns the memory footprint.